graph/adjacency/ThingAdjacency.java

```diff
@@ -21,30 +21,85 @@
 import com.vaticle.typedb.core.common.collection.KeyValue;
 import com.vaticle.typedb.core.common.iterator.FunctionalIterator;
 import com.vaticle.typedb.core.common.iterator.sorted.SortedIterator;
-import com.vaticle.typedb.core.common.iterator.sorted.SortedIterator.Order;
+import com.vaticle.typedb.core.common.iterator.sorted.SortedIterator.Seekable;
 import com.vaticle.typedb.core.graph.common.Encoding;
+import com.vaticle.typedb.core.graph.edge.Edge;
 import com.vaticle.typedb.core.graph.edge.ThingEdge;
 import com.vaticle.typedb.core.graph.iid.IID;
 import com.vaticle.typedb.core.graph.vertex.ThingVertex;
 import com.vaticle.typedb.core.graph.vertex.TypeVertex;
 
+import static com.vaticle.typedb.common.collection.Collections.list;
+import static com.vaticle.typedb.core.common.iterator.Iterators.Sorted.iterateSorted;
+
 public interface ThingAdjacency {
 
     interface In extends ThingAdjacency {
 
         InEdgeIterator edge(Encoding.Edge.Thing.Base encoding, IID... lookAhead);
 
         InEdgeIterator.Optimised edge(Encoding.Edge.Thing.Optimised encoding, TypeVertex roleType, IID... lookAhead);
+
+        @Override
+        default boolean isIn() {
+            return true;
+        }
+
+        interface InEdgeIterator {
+
+            Seekable<ThingVertex, SortedIterator.Order.Asc> from();
+
+            SortedIterator<ThingVertex, SortedIterator.Order.Asc> to();
+
+            interface Optimised extends InEdgeIterator {
+
+                Seekable<KeyValue<ThingVertex, ThingVertex>, SortedIterator.Order.Asc> fromAndOptimised();
+            }
```

Comment:
after many iterations, feels like the correct design to embed the specific types of EdgeIterators for In and Out into the In/Out `ThingAdjacency` interfaces.